### ISSUE TITLE
Could not build custom package

### ISSUE THREAD

**Issue by mdesantis** [NixOS/nix#439]
Hello! I'm new to Nix, but what I'm trying to do seems easy to me and I can't have success, so maybe it's a problem.

[env: NixOS 14.12, Nix 1.8] I want to build a package for ruby 2.2.0, so I edited Ruby 2.1.3 package expression and tried to build it, but I keep obtaining this error message:

```
error: cannot auto-call a function that has an argument without a default value (‘stdenv’)
```

This is the package I'm trying to build (with the command `nix-build ruby-2.2.0.nix` in my $HOME):

```nix
{ stdenv, fetchurl, fetchgit, fetchFromGitHub
, zlib, zlibSupport ? true
, openssl, opensslSupport ? true
, gdbm, gdbmSupport ? true
, ncurses, readline, cursesSupport ? false
, groff, docSupport ? false
, libyaml, yamlSupport ? true
, libffi, fiddleSupport ? true
, ruby_2_2_0, autoreconfHook, bison, useRailsExpress ? true
}:

let
  op = stdenv.lib.optional;
  ops = stdenv.lib.optionals;
  patchSet = import ./rvm-patchsets.nix { inherit fetchFromGitHub; };
  config = import ./config.nix fetchgit;
  baseruby = ruby_2_2_0.override { useRailsExpress = false; };
in

stdenv.mkDerivation rec {
  version = with passthru; "${majorVersion}.${minorVersion}.${teenyVersion}-p${patchLevel}";

  name = "ruby-${version}";

  src = if useRailsExpress then fetchFromGitHub {
    owner  = "ruby";
    repo   = "ruby";
    rev    = "v2_2_0";
    sha256 = "1w7rr2nq1bbw6aiagddzlrr3rl95kk33x4pv6570nm072g55ybpi";
  } else fetchurl {
    url = "http://cache.ruby-lang.org/pub/ruby/2.2/ruby-2.2.0.tar.gz";
    sha256 = "1z2092fbpc2qkv1j3yj7jdz7qwvqpxqpmcnkphpjcpgvmfaf6wbn";
  };

  # Have `configure' avoid `/usr/bin/nroff' in non-chroot builds.
  NROFF = "${groff}/bin/nroff";

  buildInputs = ops useRailsExpress [ autoreconfHook bison ]
    ++ (op fiddleSupport libffi)
    ++ (ops cursesSupport [ ncurses readline ])
    ++ (op docSupport groff)
    ++ (op zlibSupport zlib)
    ++ (op opensslSupport openssl)
    ++ (op gdbmSupport gdbm)
    ++ (op yamlSupport libyaml)
    # Looks like ruby fails to build on darwin without readline even if curses
    # support is not enabled, so add readline to the build inputs if curses
    # support is disabled (if it's enabled, we already have it) and we're
    # running on darwin
    ++ (op (!cursesSupport && stdenv.isDarwin) readline);

  enableParallelBuilding = true;

  patches = ops useRailsExpress [
    "${patchSet}/patches/ruby/2.2.0/railsexpress/01-zero-broken-tests.patch"
    "${patchSet}/patches/ruby/2.2.0/railsexpress/02-improve-gc-stats.patch"
    "${patchSet}/patches/ruby/2.2.0/railsexpress/03-display-more-detailed-stack-trace.patch"
    "${patchSet}/patches/ruby/2.2.0/railsexpress/04-backport-401c8bb.patch"
    "${patchSet}/patches/ruby/2.2.0/railsexpress/05-fix-packed-bitfield-compat-warning-for-older-gccs.patch"
  ];

  # Ruby >= 2.1.0 tries to download config.{guess,sub}
  postPatch = ''
    rm tool/config_files.rb
    cp ${config}/config.guess tool/
    cp ${config}/config.sub tool/
  '';

  configureFlags = ["--enable-shared" ]
    ++ op useRailsExpress "--with-baseruby=${baseruby}/bin/ruby"
    # on darwin, we have /usr/include/tk.h -- so the configure script detects
    # that tk is installed
    ++ ( if stdenv.isDarwin then [ "--with-out-ext=tk " ] else [ ]);

  installFlags = stdenv.lib.optionalString docSupport "install-doc";
  # Bundler tries to create this directory
  postInstall = ''
    # Bundler tries to create this directory
    mkdir -pv $out/${passthru.gemPath}
    mkdir -p $out/nix-support
    cat > $out/nix-support/setup-hook <<EOF
    addGemPath() {
      addToSearchPath GEM_PATH \$1/${passthru.gemPath}
    }

    envHooks+=(addGemPath)
    EOF
  '';

  meta = {
    license = stdenv.lib.licenses.ruby;
    homepage = "http://www.ruby-lang.org/en/";
    description = "The Ruby language";
    platforms = stdenv.lib.platforms.all;
  };

  passthru = rec {
    majorVersion = "2";
    minorVersion = "2";
    teenyVersion = "0";
    patchLevel = "0";
    libPath = "lib/ruby/${majorVersion}.${minorVersion}";
    gemPath = "lib/ruby/gems/${majorVersion}.${minorVersion}";
  };
}
```

**Comment by lethalman** [comment]
On Sat, Jan 10, 2015 at 3:48 AM, Maurizio De Santis <
notifications@github.com> wrote:

> Hello! I'm new to Nix, but what I'm trying to do seems easy to me and I
> can't have success, so maybe it's a problem.
>
Np, you still haven't clear how nix works. Try: nix-build -E '(import
<nixpkgs> {}).callPackage ./ruby-2.2.0.nix {}'

**Comment by mdesantis** [comment]
Thank you, the error is solved! \o/ I think I'm prepending nixpkgs definitions importing in this way right?

But now I get another error :-/:

```
error: anonymous function at "/home/demo/ruby-2.2.0.nix":1:1 called without required argument ‘ruby_2_2_0’, at "/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs/lib/customisation.nix":58:12
```

**Comment by lethalman** [comment]
Cause there's no ruby_2_2_0 :) Don't trial and error, understand the language. Read the nix manual, (spam) read my nix pills.

**Comment by mdesantis** [comment]
I'm doing trial and error because I can't find any resource that helps me :-/

And god if I tried! For example, in the previous case, last night I went to bed at 4am without spotting a solution... I searched `nix build custom package`, `nix import nixpgks`, I read [nix-build section in the Nix manual](http://nixos.org/nix/manual/#sec-nix-build), I read [this](http://sandervanderburg.blogspot.it/2014/07/managing-private-nix-packages-outside.html), I read even [your article :D](http://lethalman.blogspot.it/2014/11/nix-pill-16-nixpkgs-parameters.html), but I can't find any place pointing my to the solution...

So if you please would like to point me to the right resource of briefly explaining what I'm doing wrong I will be eternally grateful... I just want to avoid to reach 4am every night without spotting a solution :)

**Comment by mdesantis** [comment]
Ok, I finally managed to build a package for Ruby 2.2.0, with the help of IRC; I write here the progresses for future reference.

The error above is due to the fact that this declaration requires itself in order to provide a baseruby and apply RailsExpress patches, but since it is a custom build not declared inside all-packages.nix it can't find itself.

So I cloned nixpkgs repository, edited all-packages.nix adding `ruby_2_2_0` package, added my custom declaration as `pkgs/development/interpreters/ruby/ruby-2.2.0.nix` and run `nix-build . -A pkgs.ruby_2_2_0`.

Then I started struggling with failing sha256 hashes checks: they failed because I used the utility `sha256sum` in oder to calculate them, while has its integrity check logic and requires that the hashes of the files downloaded with `fetchurl` have to be calculated using the `nix-prefetch-url` utility, and the hashes of the files downloaded with `fetchFromGitHub` have to be calculated using the `pkgs/build-support/fetchzip/nix-prefetch-zip` script.

After that, I adjusted a couple of other things and I managed to build a brand new Ruby 2.2.0 package :cake: I even create a pull request: https://github.com/NixOS/nixpkgs/pull/5696
